src/storage/page_buffer.c

```diff
@@ -7269,59 +7305,105 @@ pgbuf_delete_from_hash_chain (THREAD_ENTRY * thread_p, PGBUF_BCB * bufptr)
     }
 }
 
+/*
+ * pgbuf_buffer_lock_enlist_thread - enlist 'this thread' as a new waiting thread at the head of the
+ *              page's buffer lock chain
+ */
+#if defined(SERVER_MODE)
+void
+pgbuf_buffer_lock_enlist_thread (THREAD_ENTRY * const thread_p, PGBUF_BUFFER_LOCK * const buffer_lock)
+{
+  thread_p->next_wait_thrd = buffer_lock->next_wait_thrd;
+  buffer_lock->next_wait_thrd = thread_p;
+}
+#endif
+
+/*
+ * pgbuf_buffer_lock_delist_thread - search 'this thread' and delist from the page's buffer lock chain
+ *
+ * NOTE: function post-conditions 'this thread' must have been found in the page's buffer lock chain
+ */
+#if defined(SERVER_MODE)
+void
+pgbuf_buffer_lock_delist_thread (THREAD_ENTRY * const thread_p, PGBUF_BUFFER_LOCK * const buffer_lock)
+{
+  THREAD_ENTRY *thrd_entry = buffer_lock->next_wait_thrd;
+  THREAD_ENTRY *prev_thrd_entry = nullptr;
+  bool thrd_found = false;
+
+  while (thrd_entry != nullptr)
+    {
+      if (thrd_entry == thread_p)
+	{
+	  if (prev_thrd_entry == nullptr)
+	    {
+	      buffer_lock->next_wait_thrd = thrd_entry->next_wait_thrd;
+	    }
+	  else
+	    {
+	      prev_thrd_entry->next_wait_thrd = thrd_entry->next_wait_thrd;
+	    }
+	  thrd_entry->next_wait_thrd = nullptr;
+	  thrd_found = true;
+	  break;
+	}
+      prev_thrd_entry = thrd_entry;
+      thrd_entry = thrd_entry->next_wait_thrd;
+    }
+
+  assert (thrd_found);
```

Comment:
Not sure this is required to be true. This kind of assertion did not exist in previous code. Please remove it if you're not 100% sure of it.

Reply:
In the only context the function is currently used in, I think the assert must hold. Unless I understood it wrongly.

Reply:
It matters who else delists the thread, and `pgbuf_unlock_page` also does it. But `pgbuf_unlock_page` also sets the resumed status, so at first glance, the assertion should stand. If you're confident, let's keep it.